### PR TITLE
Feature: add block sync by tree logic.

### PR DESCRIPTION
**Pull request by chuwen95**

### CODE REVIEW

File: bcos-sync/bcos-sync/utilities/SyncTreeTopology.cpp [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+ */
+#include "bcos-sync/bcos-sync/utilities/SyncTreeTopology.h"
+
+using namespace bcos;
+using namespace bcos::sync;
+
+void SyncTreeTopology::updateNodeInfo(bcos::crypto::NodeIDs _nodeList)
+{
+    Guard l(m_mutex);
+    if (true == bcos::crypto::KeyCompareTools::compareTwoNodeIDs(_nodeList, *m_nodeList))
```
Comment (kyonRay) [inline bcos-sync/bcos-sync/utilities/SyncTreeTopology.cpp:29]: no need to compare 'true' in if statement

File: bcos-sync/bcos-sync/utilities/SyncTreeTopology.cpp [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+    // update the nodeIndex
+    m_nodeIndex = getNodeIndexByNodeId(m_nodeList, m_nodeId);
+    updateStartAndEndIndex();
+}
+
+void SyncTreeTopology::updateAllNodeInfo(
+    bcos::crypto::NodeIDs _consensusNodes, bcos::crypto::NodeIDs _nodeList)
+{
+    Guard l(m_mutex);
+    if (true == bcos::crypto::KeyCompareTools::compareTwoNodeIDs(_nodeList, *m_nodeList) &&
```
Comment (kyonRay) [inline bcos-sync/bcos-sync/utilities/SyncTreeTopology.cpp:49]: no need to compare 'true' in if statement

File: bcos-sync/bcos-sync/utilities/SyncTreeTopology.cpp [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+    {
+        SYNCTREE_LOG(DEBUG) << LOG_DESC("getNodeIDByIndex: invalidNode")
+                            << LOG_KV("nodeIndex", _nodeIndex) << LOG_KV("nodeListSize", m_nodeNum);
+        return false;
+    }
+    _nodeID = (*m_nodeList)[_nodeIndex];
+    return true;
+}
+
+bool SyncTreeTopology::locatedInGroup()
```
Comment (kyonRay) [inline bcos-sync/bcos-sync/utilities/SyncTreeTopology.cpp:132]: inline function

File: bcos-sync/bcos-sync/utilities/SyncTreeTopology.cpp [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+    if (!locatedInGroup())
+    {
+        return;
+    }
+    // push all other consensus node to the selectedNodeList if this node is the consensus node
+    if (m_consIndex >= 0)
+    {
+        for (auto const& [idx, consNode] : *m_consensusNodes | RANGES::view::enumerate)
+        {
+            if (true == _peers->contains(consNode) &&
```
Comment (kyonRay) [inline bcos-sync/bcos-sync/utilities/SyncTreeTopology.cpp:166]: no need to compare 'true' or 'false' in if statement

File: bcos-sync/bcos-sync/BlockSync.cpp [hunk 5]
```diff
@@ -39,6 +40,13 @@ BlockSync::BlockSync(BlockSyncConfig::Ptr _config, unsigned _idleWaitMs)
     m_downloadBlockProcessor = std::make_shared<bcos::ThreadPool>("Download", 1);
     m_sendBlockProcessor = std::make_shared<bcos::ThreadPool>("SyncSend", 1);
     m_downloadingTimer = std::make_shared<Timer>(m_config->downloadTimeout(), "downloadTimer");
+
+    if (true == m_config->enableSendBlockStatusByTree())
```
Comment (kyonRay) [inline bcos-sync/bcos-sync/BlockSync.cpp:44]: no need to compare 'true' in if statement

**Review by kyonRay — APPROVED** [review]

File: bcos-sync/bcos-sync/BlockSync.cpp [hunk 6]
```diff
@@ -833,8 +850,38 @@ void BlockSync::maintainPeersConnection()
...
+    m_syncStatus->updatePeerStatus(m_config->nodeID(), statusMsg);
+    auto encodedData = statusMsg->encode();
+    BLKSYNC_LOG(TRACE) << LOG_BADGE("BlockSync") << LOG_DESC("broadcastSyncStatus")
+                       << LOG_KV("number", statusMsg->number())
+                       << LOG_KV("genesisHash", statusMsg->genesisHash().abridged())
+                       << LOG_KV("currentHash", statusMsg->hash().abridged());
+    // Note: only send status to the observers/sealers, but the OUTSIDE_GROUP node maybe
+    // observer/sealer before sync to the highest here can't use asyncSendBroadcastMessage=
+    auto const& groupNodeList =
+        m_syncTreeTopology->selectNodesForBlockSync(m_config->connectedNodeSet());
```
Comment (JimmyShi22) [inline bcos-sync/bcos-sync/BlockSync.cpp:879]: 这个接口会拿到包含轻节点的nodeid，轻节点不支持树状广播，可以用这个接口 m_config->groupNodeList

Comment (chuwen95) [reply bcos-sync/bcos-sync/BlockSync.cpp:879]: 大佬所言极是

File: bcos-tool/bcos-tool/NodeConfig.cpp [hunk 7]
```diff
@@ -618,6 +618,12 @@ void NodeConfig::loadStorageSecurityConfig(boost::property_tree::ptree const& _p
                          << LOG_KV("keyCenterUrl", storageSecurityKeyCenterUrl);
 }
 
+void NodeConfig::loadSyncConfig(const boost::property_tree::ptree &_pt)
+{
+    m_enableSendBlockStatusByTree = _pt.get<bool>("sync.enable_send_block_status_by_tree", false);
+    m_treeWidth = _pt.get<std::uint32_t>("sync.tree_width", 3);
```
Comment (JimmyShi22) [inline bcos-tool/bcos-tool/NodeConfig.cpp:624]: 同保持一致

File: bcos-sync/bcos-sync/BlockSync.cpp (hunk 6 above)
Comment (JimmyShi22) [reply bcos-sync/bcos-sync/BlockSync.cpp:879]: 补充：m_config->groupNodeList这里拿到的会有未启动的节点，还得和m_config->connectedNodeSet()做个交集

**Review by bxq2011hust — APPROVED** [review]

**Review by kyonRay — APPROVED** [review]

**Review by JimmyShi22 — APPROVED** [review]

**Pull request merged**